File: src/parser/test/driver4tstConsole_Token_Stream.cc
//----------------------------------*-C++-*----------------------------------//
/*!
 * \file   parser/test/driver4tstConsole_Token_Stream.cc
 * \author Kelly Thompson
 * \date   Wed Oct 19 14:42 2005
 * \brief  Execute the binary tstConsole_Token_Stream by redirecting the
 *         contents of console_test.inp as stdin.
 * \note   Copyright (C) 2004-1013 Los Alamos National Security, LLC.
 *         All rights reserved.
 */
//---------------------------------------------------------------------------//
// $Id$
//---------------------------------------------------------------------------//

#include "ds++/Release.hh"
#include "ds++/ScalarUnitTest.hh"
#include "ds++/path.hh"
#include <sstream>
// #include <cstdlib>
// #include <cerrno>
#include <cstring>

using namespace std;

#define PASSMSG(a) ut.passes(a)
#define ITFAILS    ut.failure(__LINE__)
#define FAILURE    ut.failure(__LINE__, __FILE__)
#define FAILMSG(a) ut.failure(a)

//---------------------------------------------------------------------------//
// In this unit test we need to check the parser's ability to accept a
// collection of tokens from standard input.  This cannot be done with a
// single unit test.  Instead, we create the binary tstConsole_Token_Stream
// that will accept data from standard input.  In this file, we actually
// execute tstConsole_Token_Stream and pipe the data from the file
// console_test.inp as input.
//---------------------------------------------------------------------------//

//---------------------------------------------------------------------------//
// TESTS
//---------------------------------------------------------------------------//

void runtest( rtt_dsxx::UnitTest &ut )
{
    // Build path for the application "tstConsole_Token_Stream"

    string tctsApp = rtt_dsxx::getFilenameComponent(
        ut.getTestPath() + std::string("tstConsole_Token_Stream"),
        rtt_dsxx::FC_NATIVE) + rtt_dsxx::exeExtension;
// #ifndef _MSC_VER    
//     // Unix needs the leading dot-slash.
//     tctsApp = std::string("./") + tctsApp;
// #endif

    // Build path for the input file "console_test.inp"
    string const ctInputFile = rtt_dsxx::getFilenameComponent(
        ut.getTestPath() + std::string("console_test.inp"),
        rtt_dsxx::FC_NATIVE);

    // String to hold command that will start the test.  For example:
    // "mpirun -np 1 ./tstConsole_Token_Stream < console_test.inp"    
    std::string consoleCommand( tctsApp + " < " + ctInputFile );

    // return code from the system command
    int errorLevel(-1);
    
    // run the test.
    errno = 0;
    errorLevel = system( consoleCommand.c_str() );
    
    // check the errorLevel
    std::ostringstream msg;
    // On Linux, the child process information is sometimes unavailable even
    // for a correct run.
    if( errorLevel == 0 || errno == ECHILD)
    {
        msg << "Successful execution of tstConsole_Token_Stream:"
            << "\n\t Standard input from: console_test.inp\n";
        PASSMSG( msg.str() );
    }
    else
    {
        msg << "Unsuccessful execution of tstConsole_Token_Stream:"
            << "\n\t Standard input from: console_test.inp\n"
            << "\t errorLevel = " << errorLevel << endl
            << "\t errno = " << errno << ", " << strerror(errno) << endl;
        FAILMSG( msg.str() );   
    }
    
    // This setup provides the possibility to parse output from each
    // run.  This potential feature is not currently implemented.            
    
    return;
}

//---------------------------------------------------------------------------//

int main(int argc, char *argv[])
{
    rtt_dsxx::ScalarUnitTest ut( argc, argv, rtt_dsxx::release );
    try { runtest(ut ); }
    UT_EPILOG(ut);
}   

//---------------------------------------------------------------------------//
// end of driver4tstConsole_Token_Stream.cc
//---------------------------------------------------------------------------//

File: src/parser/test/driver4tstConsole_Token_Stream.py
#-----------------------------*-python-*----------------------------------------#
# file   parser/test/driver4tstConsole_Token_Stream.cmake
# author Alex Long <along@lanl.gov>
# date   Wednesday, September 14, 2016, 14:16 pm
# brief  This is a python script that is used to test parser/Ipcress_Interpreter
# note   Copyright (C) 2016, Triad National Security, LLC.
#        All rights reserved.
#------------------------------------------------------------------------------#
# $Id: CMakeLists.txt 6721 2012-08-30 20:38:59Z gaber $
#------------------------------------------------------------------------------#

import sys
import re
import os # needed for Win32 check

try:
  #----------------------------------------------------------------------------#
  # Boilerplate code to get the location of the application_unit_test.py in
  # draco/config
  re_draco_config_dir = re.compile("DRACO_CONFIG_DIR=([^\s]*)")
  draco_config_dir = ""
  for arg in sys.argv:
    if (re_draco_config_dir.search(arg)):
      draco_config_dir = re_draco_config_dir.findall(arg)[0]
  if not draco_config_dir:
    raise Exception("Draco config directory not found, exiting")

  # import unit test functions
  sys.path.append(draco_config_dir)
  from application_unit_test import UnitTest
  #----------------------------------------------------------------------------#

  tDriver4tstConsole_Token_Stream = UnitTest()
  tDriver4tstConsole_Token_Stream.aut_runTests()

  ##--------------------------------------------------------------------------##
  ## Check the output
  ##--------------------------------------------------------------------------##

  print("Checking the generated output file...")

  exe_suffix = ""
  if(os.name == "nt"):
    exe_suffix = ".exe"

  # This string should be found:

  string_found = tDriver4tstConsole_Token_Stream.output_contains(
    "tstConsole_Token_Stream{0} Test: PASSED.".format(exe_suffix))
  if string_found:
    tDriver4tstConsole_Token_Stream.passmsg(
      "tstConsole_Token_Stream ran successfully.")
  else:
    tDriver4tstConsole_Token_Stream.failmsg(
      "tstConsole_Token_Stream did not run successfully.")

  ##--------------------------------------------------------------------------##
  ## Final report
  ##--------------------------------------------------------------------------##
  tDriver4tstConsole_Token_Stream.aut_report()

##----------------------------------------------------------------------------##
## Handle outstanding exceptions
##----------------------------------------------------------------------------##
except Exception:
  print("Caught exception: {0}  {1}".format( sys.exc_info()[0], \
    sys.exc_info()[1]))
  print("*****************************************************************")
  print("**** TEST FAILED.")
  print("*****************************************************************")

##----------------------------------------------------------------------------##
## End
##----------------------------------------------------------------------------##
